src/block_cache/block_tier.c

```diff
@@ -121,30 +123,44 @@ __wt_blkcache_get_handle(WT_SESSION_IMPL *session, WT_BM *bm, uint32_t objectid,
             __wt_readunlock(session, &bm->handle_array_lock);
             return (0);
         }
-
-    /* We need to add a new handle the block handle array. Upgrade to a write lock. */
     __wt_readunlock(session, &bm->handle_array_lock);
+
+    /* Open a handle for the object. */
+    WT_RET(__wt_blkcache_tiered_open(session, NULL, objectid, &new_handle));
+
+    /* We need a write lock to add a new entry to the handle array. */
     __wt_writelock(session, &bm->handle_array_lock);
 
-    /* Check to make sure the object wasn't cached while we locked. */
-    for (i = 0; i < bm->handle_array_next; ++i)
+    /* Check to see if the object was added while we opened it. */
+    for (i = 0, found = false; i < bm->handle_array_next && !found; ++i)
         if (bm->handle_array[i]->objectid == objectid) {
             *blockp = bm->handle_array[i];
-            __wt_writeunlock(session, &bm->handle_array_lock);
-            return (0);
+            found = true;
         }
 
-    /* Allocate space to store a new handle (do first for less complicated cleanup). */
-    WT_ERR(__wt_realloc_def(
-      session, &bm->handle_array_allocated, bm->handle_array_next + 1, &bm->handle_array));
+    if (!found) {
+        /* Allocate space to store the new handle and insert it in the array. */
+        WT_ERR(__wt_realloc_def(
+          session, &bm->handle_array_allocated, bm->handle_array_next + 1, &bm->handle_array));
 
-    /* Open the object */
-    WT_ERR(__wt_blkcache_tiered_open(session, NULL, objectid, blockp));
+        bm->handle_array[bm->handle_array_next++] = new_handle;
+        *blockp = new_handle;
+    }
 
-    /* Add object to block handle array. */
-    bm->handle_array[bm->handle_array_next++] = *blockp;
+    __wt_writeunlock(session, &bm->handle_array_lock);
+
+    if (found)
+        /*
+         * We raced with another thread that inserted the handle. Close ours now that we no longer
+         * hold the lock.
+         */
+        ret = __wt_bm_close_block(session, new_handle);
 
+    if (0) {
 err:
-    __wt_writeunlock(session, &bm->handle_array_lock);
+        __wt_writeunlock(session, &bm->handle_array_lock);
+        WT_TRET(__wt_bm_close_block(session, new_handle));
+    }
+
     return (ret);
```

Comment:
Can I suggest a different way to write lines 150-165?
```
err:
    __wt_writeunlock(session, &bm->handle_array_lock);
    if (found)
       WT_TRET(__wt_bm_close_block(session, new_handle));
    return (ret);
```
For me, it has the advantage of having a single unlock, at the same indentation as the lock, and done unconditionally.

Reply:
With the other change, however, we no longer have the `found` variable. Let me know what you think of the update.